drivers/gpio/gpio-litex.c

```diff
@@ -20,106 +20,309 @@
 #include <linux/platform_device.h>
 #include <linux/gpio/driver.h>
 #include <linux/gpio.h>
+#include <linux/of_irq.h>
 #include <linux/types.h>
 #include <linux/string.h>
 #include <linux/bits.h>
 #include <linux/errno.h>
+#include <linux/spinlock.h>
 #include <linux/litex.h>
 
 #define GPIO_PINS_MAX   32
+#define LITEX_GPIO_VALUE_OFFSET 0x0
+#define LITEX_GPIO_POLARITY_OFFSET 0x4
+#define LITEX_GPIO_PENDING_OFFSET 0xc
+#define LITEX_GPIO_ENABLE_OFFSET 0x10
 
 struct litex_gpio {
 	void __iomem *membase;
 	int port_direction;
 	int reg_span;
-	struct gpio_chip chip;
+	struct gpio_chip gchip;
+	struct irq_chip ichip;
+	spinlock_t gpio_lock;
+	unsigned int irq_number;
 };
 
 /* API functions */
 
-static int litex_gpio_get_value(struct gpio_chip *chip, unsigned int offset)
+static inline u32 litex_gpio_get_reg(struct litex_gpio *gpio_s, int reg_offset)
 {
-	struct litex_gpio *gpio_s = gpiochip_get_data(chip);
+	return _litex_get_reg(gpio_s->membase + reg_offset, gpio_s->reg_span);
+}
+
+static inline void litex_gpio_set_reg(struct litex_gpio *gpio_s, int reg_offset,
+				      u32 value)
+{
+	_litex_set_reg(gpio_s->membase + reg_offset, gpio_s->reg_span, value);
+}
+
+static int litex_gpio_get_value(struct gpio_chip *gchip, unsigned int offset)
+{
+	struct litex_gpio *gpio_s = gpiochip_get_data(gchip);
 	u32 regv;
 
-	if (offset >= chip->ngpio)
+	if (offset >= gchip->ngpio)
 		return -EINVAL;
 
-	regv = _litex_get_reg(gpio_s->membase, gpio_s->reg_span);
+	regv = litex_gpio_get_reg(gpio_s, LITEX_GPIO_VALUE_OFFSET);
 	return !!(regv & BIT(offset));
 }
 
-static int litex_gpio_get_multiple(struct gpio_chip *chip, unsigned long *mask,
+static int litex_gpio_get_multiple(struct gpio_chip *gchip, unsigned long *mask,
```

Comment:
I'd recomment separating cosmetic changes you feel strongly about (i.e., "it would look way cooler if instead of `chip` we used `gchip`") from actual, *functional* changes. The goal should be making it as plainly obvious and non-threatening as possible to any potential reviewer of the patch to see that either nothing functional is being changed, or that only stuff that *must* change is being changed... :smile:

Same goes for adding a wrapper around `_litex_get_reg()` -- that should be a separate (cosmetic) patch, maybe with an explanation about why it's needed, e.g. to make life easier for a future "payload" commit...

Reply:
ok, I changed the pull request to contain only the changes related to interrupt handling. I also realized that the device tree documentation needs updating. Do I need to push it in separate PR or it maybe inclueded in this ?